ratchet_core/src/handshake/server/mod.rs

```diff
@@ -311,11 +319,174 @@ where
+        .header(
+            http::header::SEC_WEBSOCKET_ACCEPT,
+            HeaderValue::try_from(sec_websocket_accept)?,
+        )
+        .header(
+            http::header::UPGRADE,
+            HeaderValue::from_static(WEBSOCKET_STR),
+        )
+        .header(
+            http::header::CONNECTION,
+            HeaderValue::from_static(UPGRADE_STR),
+        );
+
+    if let Some(subprotocol) = &subprotocol {
+        response = response.header(
+            http::header::SEC_WEBSOCKET_PROTOCOL,
+            HeaderValue::try_from(subprotocol)?,
+        );
+    }
+    if let Some(extension_header) = extension_header {
+        response = response.header(http::header::SEC_WEBSOCKET_EXTENSIONS, extension_header);
+    }
+
+    Ok(response.body(())?)
+}
+
+/// Processes a WebSocket handshake request and generates the appropriate response.
+///
+/// This function handles the server-side part of a WebSocket handshake. It parses the incoming HTTP
+/// request that seeks to upgrade the connection to WebSocket, negotiates extensions and
+/// subprotocols, and constructs an appropriate HTTP response
+/// to complete the WebSocket handshake.
+///
+/// # Arguments
+///
+/// - `request`: The incoming HTTP request from the client, which contains headers related to the
+///  WebSocket upgrade request.
+/// - `extension`: An extension that may be negotiated for the connection.
+/// - `subprotocols`: A `SubprotocolRegistry`, which will be used to attempt to negotiate a
+/// subprotocol.
+///
+/// # Returns
+///
+/// This function returns a `Result` containing:
+/// - A tuple consisting of:
+///   - An `http::Response<()>`, which represents the WebSocket handshake response.
+///     The response includes headers such as `Sec-WebSocket-Accept` to confirm the upgrade.
+///   - An optional `E::Extension`, which represents the negotiated extension, if any.
+///
+/// If the handshake fails, an `Error` is returned, which may be caused by invalid
+/// requests, issues parsing headers, or problems negotiating the WebSocket subprotocols
+/// or extensions.
+///
+/// # Type Parameters
+///
+/// - `E`: The type of the extension provider, which must implement the `ExtensionProvider`
+///   trait. This defines how WebSocket extensions (like compression) are handled.
+/// - `B`: The body type of the HTTP request. While it is discouraged for GET requests to have a body
+///  it is not technically incorrect and the use of this function is lowering the guardrails to
+///  allow for Ratchet to be more easily integrated into other libraries. It is the implementors
+///  responsibility to perform any validation on the body.
+///
+/// # Errors
+///
+/// The function returns an `Error` in cases such as:
+/// - Failure to parse the WebSocket upgrade request.
+/// - Issues building the response, such as invalid subprotocol or extension headers.
+/// - Failure to negotiate the WebSocket extensions or subprotocols.
+pub fn handshake<E, B>(
+    request: http::Request<B>,
```

Comment:
So the idea is that the caller builds a `http::Request` to pass to handshake, and then pulls the relevant stuff out of `http::Response`? I guess that's easy enough:
```rust
    let mut req_builder = http::request::Request::get(req_uri);    
    *req_builder.headers_mut().unwrap() = req_hdrs;
    let req = req_builder.body(());
```

Reply:
@nakedible-p you have a choice of how you do it:

- You can call `parse_request` and it will attempt to parse it as a valid upgrade request and then it will give you a `UpgradeRequest` which contains the relevant parts and then you can call `build_response` to derive the WebSocket response
- You can either just pass `handshake` a request to attempt to upgrade and then it will give you a response which you can pass to the client and then you can build your WebSocket using `WebSocket::from_upgraded` - this is just shorthand for the above two functions.

Reply:
I more meant that the input isn't just a `HeaderMap`, so the caller needs to build a `Request`, and the response isn't just a `HeaderMap` so the caller needs to call `response.into_parts()` to get the headers. But like said, easy enough.

Reply:
In order to parse the request it needs the version, method and headers. For the response it sets the version, status and headers, and correctly types the response to have no HTTP body. If you're interested, I've done an initial pass at upgrading a Hyper web server to use this extracted functionality: https://github.com/swimos/swim-rust/pull/705/files